src/Build/Logging/FancyLogger/FancyLoggerNode.cs

```diff
@@ -4,148 +4,136 @@
+        public FancyLoggerProjectNode(ProjectStartedEventArgs args)
         {
-            Id = id;
+            Id = args.ProjectId;
+            ProjectPath = args.ProjectFile!;
+            Finished = false;
+            FinishedTargets = 0;
         }
-        public FancyLoggerNode(int id, FancyLoggerNodeType type)
+
+        public void Log()
         {
-            switch (type)
+            string lineContents = ANSIBuilder.Alignment.SpaceBetween(
+                $"{(Finished ? ANSIBuilder.Formatting.Color("✓", ANSIBuilder.Formatting.ForegroundColor.Green) : ANSIBuilder.Graphics.Spinner())} {ANSIBuilder.Formatting.Dim("Project: ")} {ANSIBuilder.Formatting.Color(ANSIBuilder.Formatting.Bold(GetUnambiguousPath(ProjectPath)), Finished ? ANSIBuilder.Formatting.ForegroundColor.Green : ANSIBuilder.Formatting.ForegroundColor.Default )}",
+                $"({FinishedTargets} targets completed)",
+                Console.WindowWidth
+            );
+            // Create or update line
+            if (Line == null) Line = FancyLoggerBuffer.WriteNewLine(lineContents);
+            else FancyLoggerBuffer.UpdateLine(Line.Id, lineContents);
+            // If current target
+            if (CurrentTargetNode == null) return;
+            // Create or update
+            if (Finished && CurrentTargetLine != null)
             {
-                case FancyLoggerNodeType.Build:
-                    Id = $"build-{id}";
-                    break;
-                case FancyLoggerNodeType.Project:
-                    Id = $"project-{id}";
-                    break;
-                case FancyLoggerNodeType.Target:
-                    Id = $"target-{id}";
-                    break;
-                case FancyLoggerNodeType.Task:
-                    Id = $"task-{id}";
-                    break;
-                case FancyLoggerNodeType.Message:
-                    Id = $"message-{id}";
-                    break;
-                case FancyLoggerNodeType.Warning:
-                    Id = $"warning-{id}";
-                    break;
-                case FancyLoggerNodeType.Error:
-                    Id = $"error-{id}";
-                    break;
-                default:
-                    Id = id.ToString(); break;
+                FancyLoggerBuffer.DeleteLine(CurrentTargetLine.Id);
+                return;
             }
+            string currentTargetLineContents = $"   └── {CurrentTargetNode.TargetName} : {CurrentTargetNode.CurrentTaskNode?.TaskName ?? String.Empty}";
+            if (CurrentTargetLine == null) CurrentTargetLine = FancyLoggerBuffer.WriteNewLineAfter(currentTargetLineContents, Line.Id);
+            else FancyLoggerBuffer.UpdateLine(CurrentTargetLine.Id, currentTargetLineContents);
         }
-        public FancyLoggerNode? Find(string id)
+
+        public void AddTarget(TargetStartedEventArgs args)
+        {
+            CurrentTargetNode = new FancyLoggerTargetNode(args);
```

Comment:
This seems to assume that there is only ever one target executing at a time, and once we've started executing a target, we execute it to completion and never look at it again.

That's often true but not guaranteed. As a simple example of when it fails, we have a task called CallTarget. You can execute that within Target Foo to go execute Target Bar. Afterwards, it will return to execute the rest of Foo.

With that in mind, might it be necessary to have a list of targets currently executing?

Reply:
This is true but may not be super important at the moment. If we track most-recently-started target, in the `CallTarget` case it'll be:

| Time | Target Running              | Target logged               |
|------|-----------------------------|-----------------------------|
| 1    | `TargetThatHasCallTargetInIt` | `TargetThatHasCallTargetInIt` |
| 2    | `TargetThatWasCalled`         | `TargetThatWasCalled`         |
| 3    | `TargetThatHasCallTargetInIt` | `TargetThatWasCalled`         |

The 1->2 transition is fine. The 2->3 transition is potentially a problem if there's a long-running task after the `CallTarget`, since the user could misattribute the problem. I don't think that's super common. We could also potentially work around it by updating the `Target` state within `AddTask`--if we start a new task, make sure the target containing that task is on the top of the list.